### DISCUSSION TITLE
Is it possible to make auto increment id not nullable?

### DISCUSSION THREAD

**Discussion by barbalex** [isar/isar#88] [Help] [answered]
Being a flutter/dart noob coming from JavaScript I am fighting with type soundness.

The docs say that an auto incrementing id can be created like this:
```dart
int? id;
```

I am using a slight variation of it because my data structure has the id field reserved for a uuid due to distributed offline capability in my `project` class:
```dart
@Id()
int? isarId;

@Index()
late String? id;
```

So I assume that the `isarId` field is an auto increment id that is set by isar. I also assume that it is always set and there will never be a null value.

Assume that somewhere in my code I want to replace all projects with fresh data fetched from the server:
```dart
await isar.writeTxn((isar) async {
  List<int?> ids = await isar.projects.where().isarIdProperty().findAll();
  await isar.projects.deleteAll(ids);
  await isar.projects.putAll(projects);
});
```

My problem is that `await isar.projects.deleteAll(ids);` will not work due to this compile time error: `The argument type 'List<int?>' can't be assigned to the parameter type 'List<int>'`.

Casting `ids` to `List<int>` leads to a runtime error `_CastError (type 'CastList<int?, int?>' is not a subtype of type 'List<int>' in type cast)`.

I also tried to declare the id late in the project class:
```dart
@Id()
late int isarId;

@Index()
late String? id;
```
This leads to an error `LateError (LateInitializationError: Field 'isarId' has not been initialized.)` at runtime.

So I have no idea how to do this. Surely there is an easy solution but being a typing noob I have missed it.

**Comment by simc** [comment, marked as answer]
@barbalex You are right this is an issue currently but I'm already working on a solution. Will be part of the next release 👍

It will look something like this:
```dart
@Collection()
class Person {
  int id = Isar.AUTO_INCREMENT_ID;

  String name?;
}
```
